cpp/dolfinx/refinement/ParallelRefinement.cpp

```diff
@@ -160,90 +213,43 @@ void ParallelRefinement::update_logical_edgefunction()
-  std::vector<std::int32_t> vertex_to_x(num_vertices);
-  auto map_c = _mesh.topology().index_map(tdim);
-  assert(map_c);
-  for (int c = 0; c < map_c->size_local() + map_c->num_ghosts(); ++c)
-  {
-    auto vertices = c_to_v->links(c);
-    auto dofs = x_dofmap.links(c);
-    for (int i = 0; i < vertices.rows(); ++i)
-    {
-      // FIXME: We are making an assumption here on the
-      // ElementDofLayout. We should use an ElementDofLayout to map
-      // between local vertex index an x dof index.
-      vertex_to_x[vertices[i]] = dofs(i);
-    }
-  }
-
-  // Copy over existing mesh vertices
-  const Eigen::Array<double, Eigen::Dynamic, 3, Eigen::RowMajor>& x_g
-      = _mesh.geometry().x();
-
-  std::int32_t num_new_vertices
-      = std::count(_marked_edges.begin(), _marked_edges.end(), true);
-
-  _new_vertex_coordinates.resize(num_vertices + num_new_vertices, 3);
-
-  for (int v = 0; v < num_vertices; ++v)
-    _new_vertex_coordinates.row(v) = x_g.row(vertex_to_x[v]);
-
-  // Compute all edge mid-points
-  auto map_e = _mesh.topology().index_map(1);
-  assert(map_e);
-  const std::int32_t num_edges = map_e->size_local() + map_e->num_ghosts();
-  Eigen::Array<int, Eigen::Dynamic, 1> edges(num_edges);
-  std::iota(edges.data(), edges.data() + edges.rows(), 0);
-  const Eigen::Array<double, Eigen::Dynamic, 3, Eigen::RowMajor> midpoints
-      = mesh::midpoints(_mesh, 1, edges);
-
-  // Tally up unshared marked edges, and shared marked edges which are
-  // owned on this process. Index them sequentially from zero.
-  // Locally owned edges
-
-  std::int64_t n = 0;
+  // Add new edge midpoints to list of vertices
+  int n = 0;
+  std::map<std::int32_t, std::int64_t> local_edge_to_new_vertex;
   for (int local_i = 0; local_i < edge_index_map->size_local(); ++local_i)
   {
     if (_marked_edges[local_i] == true)
     {
-      _new_vertex_coordinates.row(n + num_vertices) = midpoints.row(local_i);
-      auto it = _local_edge_to_new_vertex.insert({local_i, n});
+      auto it = local_edge_to_new_vertex.insert({local_i, n});
       assert(it.second);
       ++n;
     }
   }
-
-  // Calculate global range for new local vertices
+  const int num_new_vertices = n;
   const std::size_t global_offset
       = MPI::global_offset(_mesh.mpi_comm(), num_new_vertices, true)
-        + _mesh.topology().index_map(0)->size_global();
+        + _mesh.topology().index_map(0)->local_range()[1];
+
+  for (auto& e : local_edge_to_new_vertex)
+    e.second += global_offset;
+
+  // Create actual points
+  _new_vertex_coordinates
+      = create_new_geometry(_mesh, local_edge_to_new_vertex);
 
   // If they are shared, then the new global vertex index needs to be
-  // sent off-process.  Add offset to map, and collect up any shared new
-  // vertices that need to send the new index off-process
+  // sent off-process.
 
   int num_neighbours = _marked_for_update.size();
   std::vector<std::vector<std::int64_t>> values_to_send(num_neighbours);
```

Comment:
Can we avoid `std::vector<std::vector>>`?

Reply:
Possible by traversing the data twice, but makes the code longer and more complex, and I am not sure it is more efficient. The number of vectors here is quite limited, as it is `num_neighbours` not `mpi_size`.